### DISCUSSION TITLE
Contributing: How to add cluster command handler

### DISCUSSION THREAD

**Discussion by jcam** [zigpy/zigpy#677] [Q&A]
Hi! I'm working on adding additional lock management and notification support to zigpy and ZHA in HA.
Related, I've come across this error:
"No handler for cluster command"

Where would I go to create a cluster command? Got any examples of a PR where one was added so I have somewhere to start?

Thank you!!

**Comment by Adminiuga** [comment]
Depends whether you need to handle the cluster command at "device level" or "application level". In 99% that would be the "application" level and those you handle in the "cluster listeners" attached to the cluster. the "cluster listener" may have a `def cluster_command(self, tsn: int, command_id: int, args: List[Any])` sync function to handle those.
In home assistant, ZHA core channel are the "cluster listeners". An example would be https://github.com/home-assistant/core/blob/3ebd5aff98fa2bc4a910ee4caea02af148dd6571/homeassistant/components/zha/core/channels/general.py#L223

**Reply by jcam** [reply]
I did manage to figure this out (it is the operation_event_notification piece of [the zha lock code PR](https://github.com/home-assistant/core/pull/47208)), but i see in debug logs that it is still saying "No handler for cluster command" even though the app is now handling it.
Should I safely ignore this, or should I add an empty handler on the device side?

**Reply by Adminiuga** [reply]
Yep, that one you 🥫 ignore
